oauth2-server/src/main/java/com/clouway/oauth2/AuthCodeAuthorization.java

```diff
@@ -20,33 +22,38 @@
   private final ClientAuthorizationRepository clientAuthorizationRepository;
   private final IdentityFinder identityFinder;
   private final AuthorizedClientActivity authorizedClientActivity;
+  private final CodeVerifier codeVerifier;
 
-  AuthCodeAuthorization(ClientAuthorizationRepository clientAuthorizationRepository, IdentityFinder identityFinder, AuthorizedClientActivity authorizedClientActivity) {
+  AuthCodeAuthorization(ClientAuthorizationRepository clientAuthorizationRepository, IdentityFinder identityFinder, AuthorizedClientActivity authorizedClientActivity, CodeVerifier codeVerifier) {
     this.clientAuthorizationRepository = clientAuthorizationRepository;
     this.identityFinder = identityFinder;
     this.authorizedClientActivity = authorizedClientActivity;
+    this.codeVerifier = codeVerifier;
```

Comment:
You can move codeVerification check one level up to the callstack, e.g in another implementation of ClientActivity. This will remove this new dependency here and will make tests clear for this part. 

```java
// e.g code exchange guards calling of AuthCodeAuthorization 
new CodeExchangeVerificationFlow(codeVerifier, new AuthCodeAuthorization()) 
```

Reply:
Wouldn't this duplicate the roundtrip to the database for Authorization object ? We will check if the codeChallenge stored with the Authorization matches then in the AuthCodeAuthorization we extract the authorization again. Thinking for a solution.

Reply:
Or I guess we can do the authorization first and if the verification passes then we will return it.

Reply:
It will perform just existence and syntax check. 

The first 2 if statements which are turning OAuth2Error in case of match.

Reply:
Yeah, it may force you to add a new interface which has execute method that accepts Authorization.

Reply:
Added the new interface. I'm not very sure if this doesn't make things more complex than they should be (for a token to be issues 3 activities are passed) but the dependencies have lowered and the tests are clearer so we have that going for us.